File: c++/Woied/Woied/Woeid.cpp
#include "Woeid.h"
#include <cstdarg>
#include <memory>

Woeid::Woeid(long id_, std::string city_, std::string country_, float latitude_, float longitude_)
{
	id = id_;
	city = city_;
	country = country_;
	latitude = latitude_;
	longitude = longitude_;
}

Woeid::Woeid(const Woeid & other) 
{
	id = other.id;
	city = other.city;
	country = other.country;
	latitude = other.latitude;
	longitude = other.longitude;
}

long Woeid::getID() 
{
	return id;
}

void Woeid::setID(long id_) 
{
	id = id_;
}

std::string Woeid::getCity() 
{
	return city;
}

void Woeid::setCity(std::string city_) 
{
	city = city_;
}

std::string Woeid::getCountry() 
{
	return country;
}

void Woeid::setCountry(std::string country_) 
{
	country = country_;
}

float Woeid::getLatitude() 
{
	return latitude;
}

void Woeid::setLatitude(float latitude_) 
{
	latitude = latitude_;
}

float Woeid::getLongitude() 
{
	return longitude;
}

void Woeid::setLongitude(float longitude_) 
{
	longitude = longitude_;
}

// https://rextester.com/WMK79392
std::string format(const char* format, ...)
{
	va_list args;
	va_start(args, format);
#ifdef _MSC_VER
	size_t size = std::snprintf(nullptr, 0, format, args) + 1; // Extra space for '\0'
	std::unique_ptr<char[]> buf(new char[size]);
	std::vsnprintf(buf.get(), size, format, args);
	return std::string(buf.get(), buf.get() + size - 1); // We don't want the '\0' inside
#else
	int size = _vscprintf(format, args);
	std::string result(++size, 0);
	vsnprintf_s((char*)result.data(), size, _TRUNCATE, format, args);
	return result;
#endif
	va_end(args);
}

std::string Woeid::toString()
{
	char temp[256];
	sprintf_s(temp, "%ld,%s,%s,%f,%f", id, city.c_str(), country.c_str(), latitude, longitude);
	std::string t = temp;
	return t;
	//return format("%ld,%s,%s,%f,%f", id, city.c_str(), country.c_str(), latitude, longitude);
}

File: java/Woeid/TCP/server/Woeid.java
import java.util.InputMismatchException; 
import java.util.UnknownFormatConversionException; 

public class Woeid implements Comparable<Woeid> { 
	private long id; 
	private String city;
	private String country;
	private double latitude;
	private double longitude;		

	public Woeid() { 
		this(-1, "", "", 0.0, 0.0);
	}

	public Woeid(long id, String city, String country, double latitude, double longitude) {
		this.id = id;
		this.city = city;
		this.country = country;
		this.latitude = latitude;
		this.longitude = longitude;
	}

    
	public Woeid(Woeid other) { 
		this(other.id, other.city, other.country, other.latitude, other.longitude);
	}
	

	public long getID() {
		return this.id;
	}

	public void setID(long id) {
		this.id = id;
	}
	
	public String getCity() {
		return this.city;
	}

	public void setCity(String city) {
		this.city = city;
	}

	public String getCountry() {
		return this.country;
	}

	public void setCountry(String country) {
		this.country = country;
	}
	
	public double getLatitude() {
		return this.latitude;
	}

	public void setLatitude(double latitude) {
		this.latitude = latitude;
	}
	
	public double getLongitude() {
		return this.longitude;
	}

	public void setLongitude(double longitude) {
		this.longitude = longitude;
	}

	@Override
	public String toString() {
		return String.format("%d,%s,%s,%f,%f", this.id, this.city, this.country, this.latitude, this.longitude);
	}
	
	
	@Override
	public int compareTo(Woeid other) { 
    	return (int) (this.id - other.id);
	}
	    
  
	public static Woeid getData(String[] items) {
		if (items.length != 5) 
			return null;
		try {
			//System.out.println("items : " + Arrays.toString(items));
			long id = Long.parseLong(items[0]);
			String city = items[1];
			String country = items[2];
			double latitude = Double.parseDouble(items[3]);
			double longitude = Double.parseDouble(items[4]);
			return new Woeid(id, city, country, latitude, longitude);
		} catch (InputMismatchException e) {
			System.out.println("InputMismatchException!");
		} catch (UnknownFormatConversionException e) {
			System.out.println("UnknownFormatConversionException!");
		}
		return null;
	}
    // csv String -> Woeid, line
	public static Woeid getData(String line) {
		String[] items = line.split(",",5); 
		return Woeid.getData(items); // String -> Woeid
	}
}